torch/nn/modules/batchnorm.py

```diff
@@ -3,12 +3,15 @@
 from torch.nn.parameter import Parameter
 from .. import functional as F
 from .. import init
+from ..._jit_internal import weak_module, weak_script_method
 
 
 # TODO: check contiguous in THNN
 # TODO: use separate backend functions?
+@weak_module
 class _BatchNorm(Module):
     _version = 2
+    __constants__ = ['training', 'track_running_stats', 'momentum']
```

Comment:
if `training` is a constant, then `model.eval()` will fail right?
Another reported issue about `training` being constant (for dropout): https://github.com/pytorch/pytorch/issues/13991

Reply:
Yes this is a known issue for us, I also documented this issue in #12894, I think we still don't hav e a good solution yet, that's why we go with the weak_module approach to not interfere with users calling from python.

Reply:
@soumith to be clear on this, this will only fail when users using JIT and making a call from `ScriptModule` then want to set `model.eval()`, it will not affect the python behaviors.

Reply:
i see. thanks for the clarification.

So you are saying `model.eval()` will work on a ScriptModule correctly?
Or are you saying `model.eval()` on a regular nn.Module will continue to work correctly as before?

Reply:
I mean the second one :). 

To make `model.eval()` work on `ScriptModule` requires more changes on our current infrastructure, and we might want to work on this after the standard library export.